src/arithmetic/rewrite_simplify.cc

```diff
@@ -0,0 +1,646 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file rewrite_simplify.cc
+ * \brief Rewrite-rule based simplification.
+ */
+// Acknowledgement: Most rewrite-rules are from Halide.
+#include <tvm/arithmetic.h>
+#include <tvm/expr_operator.h>
+#include <tvm/ir_mutator.h>
+#include "const_fold.h"
+#include "pattern_match.h"
+
+namespace tvm {
+namespace arith {
+
+using namespace ir;
+
+// macro for doing simple rewrite
+#define TVM_TRY_REWRITE(SrcExpr, ResExpr)       \
+  if ((SrcExpr).Match(ret)) {                   \
+    return (ResExpr).Eval();                    \
+  }
+
+// macro for rewrite + recursively rewrite ResExpr
+#define TVM_TRY_RECURSIVE_REWRITE(SrcExpr, ResExpr) \
+  if ((SrcExpr).Match(ret)) {                       \
+    return RecursiveRewrite((ResExpr).Eval());      \
+  }
+
+// macro rewrite only if CondExor is true after match.
+#define TVM_TRY_REWRITE_IF(SrcExpr, ResExpr, CondExpr)  \
+  if ((SrcExpr).Match(ret) && (CondExpr)) {             \
+    return (ResExpr).Eval();                            \
+  }
+
+// macro rewrite + recursive_rewrite only if CondExor is true after match.
+#define TVM_TRY_RECURSIVE_REWRITE_IF(SrcExpr, ResExpr, CondExpr)  \
+  if ((SrcExpr).Match(ret) && (CondExpr)) {                       \
+    return RecursiveRewrite((ResExpr).Eval());                    \
+  }
+
+
+// NOTE for developers:
+//
+// We mainly focus on index expression simplification.
+// Besides the RewriteSimplifier, some cases can be better
+// handled by CanonicalSimplifier.
+//
+class RewriteSimplifier::Impl : public IRMutator {
+ public:
+  explicit Impl(Analyzer* parent)
+      : parent_(parent) {}
+
+  void Update(const Var& var,
+              const Expr& info,
+              bool override) {
+    if (!override) {
+      CHECK(!var_map_.count(var));
+    }
+    var_map_[var] = info;
+  }
+
+  // Run simplification in post order
+  Expr PostOrderSimplify(Expr expr, int max_iter = 2) {
```

Comment:
const Expr& for consistency and efficency

Reply:
Need to mutate expr during the process, so keep it non-const, will change the RecursiveRewrite one